// generator-jhipster/dist/generators/base/command.js
import { parseCreationTimestamp } from "./support/timestamp.js";
const command = {
    configs: {
        useVersionPlaceholders: {
            description: 'replace mutable versions with placeholders',
            cli: {
                type: Boolean,
                env: 'VERSION_PLACEHOLDERS',
                hide: true,
            },
            scope: 'generator',
        },
        skipChecks: {
            description: 'Check the status of the required tools',
            cli: {
                type: Boolean,
            },
            scope: 'generator',
        },
        experimental: {
            description: 'Enable experimental features. Please note that these features may be unstable and may undergo breaking changes at any time',
            cli: {
                type: Boolean,
            },
            scope: 'generator',
        },
        blueprints: {
            description: 'A comma separated list of one or more generator blueprints to use for the sub generators, e.g. --blueprints kotlin,vuejs',
            cli: {
                type: String,
            },
            scope: 'none',
        },
        disableBlueprints: {
            description: 'Disable blueprints support',
            cli: {
                type: Boolean,
            },
            scope: 'generator',
        },
        debugEnabled: {
            description: 'Enable debugger',
            cli: {
                name: 'debug',
                type: Boolean,
                alias: 'd',
            },
            scope: 'generator',
        },
        reproducible: {
            description: 'Try to reproduce changelog',
            cli: {
                type: Boolean,
            },
            scope: 'none',
        },
        ignoreNeedlesError: {
            description: 'Ignore needles failures',
            cli: {
                type: Boolean,
                hide: true,
            },
            scope: 'generator',
        },
        nodePackageManager: {
            description: 'Force an unsupported client package manager',
            cli: {
                type: String,
            },
            scope: 'storage',
        },
        ignoreErrors: {
            description: "Don't fail on prettier errors.",
            cli: {
                type: Boolean,
            },
            scope: 'none',
        },
        creationTimestamp: {
            description: 'Project creation timestamp (used for reproducible builds)',
            cli: {
                type: parseCreationTimestamp,
            },
            scope: 'storage',
        },
        blueprint: {
            description: 'DEPRECATED: Specify a generator blueprint to use for the sub generators',
            internal: {
                type: Array,
            },
            scope: 'none',
        },
        jdlDefinition: {
            internal: {
                type: Object,
            },
            scope: 'none',
        },
    },
};
export default command;
