// generator-jhipster/dist/generators/liquibase/command.js
const command = {
    configs: {
        entities: {
            argument: {
                description: 'Which entities to generate a new changelog',
                type: Array,
            },
            scope: 'none',
        },
        skipFakeData: {
            description: 'Skip generation of fake data for development',
            cli: {
                type: Boolean,
            },
            scope: 'storage',
        },
        recreateInitialChangelog: {
            description: 'Recreate the initial database changelog based on the current config',
            cli: {
                type: Boolean,
            },
            scope: 'generator',
        },
        skipDbChangelog: {
            description: 'Skip the generation of database migrations',
            cli: {
                type: Boolean,
            },
            scope: 'none',
        },
        incrementalChangelog: {
            cli: {
                description: 'Creates incremental database changelogs',
                type: Boolean,
            },
            jdl: {
                type: 'boolean',
                tokenType: 'BOOLEAN',
            },
            scope: 'storage',
        },
    },
};
export default command;
