// generator-jhipster/dist/generators/spring-boot/application.js
import { getDatabaseTypeData } from "../server/support/database.js";
export const mutateApplicationLoading = {
    __override__: false,
    prodActiveProfiles: () => [],
    devActiveProfiles: () => [],
};
export const mutateApplicationPreparing = {
    __override__: false,
    springDataDescription: ({ databaseType, reactive }) => {
        let springDataDatabase;
        if (databaseType !== 'sql') {
            springDataDatabase = getDatabaseTypeData(databaseType).name;
            if (reactive) {
                springDataDatabase += ' reactive';
            }
        }
        else {
            springDataDatabase = reactive ? 'R2DBC' : 'JPA';
        }
        return `Spring Data ${springDataDatabase}`;
    },
};
export const mutateFilterableField = {
    __override__: false,
    propertyJavaFilterName: ({ fieldName }) => fieldName,
    propertyJavaFilterJavaBeanName: ({ fieldInJavaBeanMethod }) => fieldInJavaBeanMethod,
    propertyFilterConsumerName: ({ propertyJavaFilterJavaBeanName }) => `set${propertyJavaFilterJavaBeanName}`,
    propertyFilterSupplierName: ({ propertyJavaFilterJavaBeanName }) => `get${propertyJavaFilterJavaBeanName}`,
};
export const mutateFilterableRelationship = {
    __override__: false,
    propertyJavaFilterName: ({ relationshipName }) => `${relationshipName}Id`,
    propertyJavaFilterJavaBeanName: ({ relationshipNameCapitalized }) => `${relationshipNameCapitalized}Id`,
    propertyFilterConsumerName: ({ propertyJavaFilterJavaBeanName }) => `set${propertyJavaFilterJavaBeanName}`,
    propertyFilterSupplierName: ({ propertyJavaFilterJavaBeanName }) => `get${propertyJavaFilterJavaBeanName}`,
};
