// generator-jhipster/dist/generators/app/command.js
const command = {
    configs: {
        defaults: {
            cli: {
                description: 'Execute jhipster with default config',
                type: Boolean,
            },
            scope: 'none',
        },
        jhiPrefix: {
            cli: {
                description: 'Add prefix before services, controllers and states name',
                type: String,
            },
            scope: 'storage',
        },
        entitySuffix: {
            cli: {
                description: 'Add suffix after entities name',
                type: String,
            },
            scope: 'storage',
        },
        dtoSuffix: {
            cli: {
                description: 'Add suffix after dtos name',
                type: String,
            },
            scope: 'storage',
        },
        pkType: {
            cli: {
                description: 'Default primary key type (beta)',
                type: String,
            },
            scope: 'storage',
        },
        testFrameworks: {
            description: 'Test frameworks to be generated',
            cli: {
                type: Array,
            },
            configure(gen, value) {
                if (value) {
                    gen.jhipsterConfig.testFrameworks = [...new Set([...(gen.jhipsterConfig.testFrameworks || []), ...value])];
                }
            },
            scope: 'none',
        },
    },
    import: [
        'base',
        'bootstrap',
        'jhipster:base-application:bootstrap',
        'jhipster:jdl:bootstrap',
        'common',
        'server',
        'client',
        'git',
        'cypress',
        'languages',
    ],
};
export default command;
