// generator-jhipster/dist/generators/java/support/prepare-entity.js
import path from 'node:path';
import { mutateData, normalizePathEnd } from "../../../lib/utils/index.js";
import { formatDocAsApiDescription, formatDocAsJavaDoc } from "./doc.js";
export function prepareEntity(entity, application) {
    const { packageFolder, packageName } = application;
    mutateData(entity, {
        __override__: false,
        entityJavaPackageFolder: ({ entityPackage }) => (entityPackage ? `${entityPackage.replace(/\./g, '/')}/` : ''),
        entityAbsoluteFolder: ({ entityJavaPackageFolder }) => normalizePathEnd(path.join(packageFolder, entityJavaPackageFolder)),
        entityAbsolutePackage: ({ entityPackage }) => (entityPackage ? [packageName, entityPackage].join('.') : packageName),
        entityAbsoluteClass: ({ entityAbsolutePackage, persistClass }) => `${entityAbsolutePackage}.domain.${persistClass}`,
        entityJavadoc: ({ documentation }) => (documentation ? formatDocAsJavaDoc(documentation) : documentation),
        entityApiDescription: ({ documentation }) => (documentation ? formatDocAsApiDescription(documentation) : documentation),
        importApiModelProperty: ({ relationships, fields }) => relationships.some(relationship => relationship.documentation) || fields.some(field => field.documentation),
    });
}
