// generator-jhipster/dist/generators/languages/command.js
import detectLanguage from "./support/detect-language.js";
const command = {
    arguments: {
        languages: {
            description: 'Languages to generate',
            type: Array,
            required: false,
        },
    },
    configs: {
        languages: {
            cli: {
                type: Array,
                hide: true,
            },
            scope: 'storage',
        },
        languagesDefinition: {
            cli: {
                type: Array,
                hide: true,
            },
            scope: 'storage',
        },
        enableTranslation: {
            cli: {
                description: 'Enable translation',
                type: Boolean,
            },
            scope: 'storage',
        },
        language: {
            cli: {
                alias: 'l',
                description: 'Language to be added to application (existing languages are not removed)',
                type: Array,
            },
            configure(gen, value) {
                if (value) {
                    gen.jhipsterConfig.languages = [...(gen.jhipsterConfig.languages ?? []), ...value];
                }
            },
            scope: 'none',
        },
        nativeLanguage: {
            cli: {
                alias: 'n',
                description: 'Set application native language',
                type: String,
                required: false,
            },
            configure(gen, value) {
                if (value) {
                    if (value === true || value === 'true') {
                        gen.jhipsterConfig.nativeLanguage = detectLanguage();
                    }
                    else if (typeof value === 'string') {
                        gen.jhipsterConfig.nativeLanguage = value;
                    }
                    if (!gen.jhipsterConfig.languages) {
                        gen.jhipsterConfig.languages = [gen.jhipsterConfig.nativeLanguage];
                    }
                }
            },
            scope: 'storage',
        },
        regenerateLanguages: {
            cli: {
                description: 'Regenerate languages',
                type: Boolean,
            },
            scope: 'generator',
        },
    },
};
export default command;
