// generator-jhipster/dist/generators/bootstrap/command.js
const command = {
    configs: {
        autoCrlf: {
            description: 'Detect line endings',
            cli: {
                type: Boolean,
            },
            scope: 'storage',
        },
        removeNeedles: {
            description: 'Remove needles, needles are used by entity and language generators to inject code',
            cli: {
                type: Boolean,
            },
            jdl: {
                tokenType: 'BOOLEAN',
                type: 'boolean',
            },
            scope: 'storage',
        },
        skipPrettier: {
            description: 'Skip prettier transformation',
            cli: {
                type: Boolean,
                hide: true,
            },
            scope: 'generator',
        },
        skipEslint: {
            description: 'Skip ESlint transformation',
            cli: {
                type: Boolean,
                hide: true,
            },
            scope: 'generator',
        },
        refreshOnCommit: {
            description: 'Refresh files on commit',
            cli: {
                type: Boolean,
                hide: true,
            },
            scope: 'generator',
        },
    },
};
export default command;
