// generator-jhipster/dist/generators/java-simple-application/command.js
const command = {
    configs: {
        withGeneratedFlag: {
            description: 'Add a GeneratedByJHipster annotation to all generated java classes and interfaces',
            cli: {
                type: Boolean,
            },
            scope: 'storage',
        },
        packageInfoFile: {
            description: 'write package-info.java file for every package',
            cli: {
                type: Boolean,
                default: true,
                hide: true,
            },
            scope: 'generator',
        },
        packageName: {
            cli: {
                type: String,
            },
            prompt: (gen) => ({
                type: 'input',
                message: 'What is your default Java package name?',
                default: gen.jhipsterConfigWithDefaults.packageName,
                validate: (input) => /^([a-z_][a-z0-9_]*(\.[a-z_][a-z0-9_]*)*)$/.test(input)
                    ? true
                    : 'The package name you have provided is not a valid Java package name.',
            }),
            scope: 'storage',
            description: 'The package name for the generated application',
        },
        packageFolder: {
            cli: {
                type: String,
                hide: true,
            },
            scope: 'storage',
        },
        projectVersion: {
            description: 'project version to use, this option is not persisted',
            cli: {
                type: String,
                env: 'JHI_PROJECT_VERSION',
            },
            scope: 'storage',
        },
        jhipsterDependenciesVersion: {
            description: 'jhipster-dependencies version to use, this option is not persisted',
            cli: {
                type: String,
                env: 'JHIPSTER_DEPENDENCIES_VERSION',
            },
            scope: 'generator',
        },
    },
    import: ['jhipster:java-simple-application:build-tool'],
};
export default command;
