// generator-jhipster/dist/generators/base-application/internal/types/field-types.js
import { blobFieldTypesValues } from "../../../../lib/jhipster/field-types.js";
export const isBlobType = (fieldType) => Object.values(blobFieldTypesValues).includes(fieldType);
export const getBlobContentType = (fieldType) => {
    switch (fieldType) {
        case 'AnyBlob': {
            return 'any';
        }
        case 'ImageBlob': {
            return 'image';
        }
        case 'TextBlob': {
            return 'text';
        }
        default: {
            return undefined;
        }
    }
};
export const isFieldBlobType = (field) => isBlobType(field.fieldType);
export const isFieldBinaryType = (field) => isBlobType(field.fieldType) || field.fieldType === 'byte[]';
export const isFieldEnumType = (field) => Boolean(field.fieldValues);
export const isFieldNotEnumType = (field) => !field.fieldValues;
export function convertFieldBlobType(field) {
    // Convert fieldTypes to correct fieldTypes
    switch (field.fieldTypeBlobContent) {
        case 'image': {
            field.fieldType = 'ImageBlob';
            field.fieldTypeBlobContent = undefined;
            break;
        }
        case 'any': {
            field.fieldType = 'AnyBlob';
            field.fieldTypeBlobContent = undefined;
            break;
        }
        case 'text': {
            field.fieldType = 'TextBlob';
            field.fieldTypeBlobContent = undefined;
            break;
        }
        default: {
            // Unknown type is not supported.
            // Fallback to ByteBuffer for cassandra databases.
            field.fieldType = 'ByteBuffer';
            break;
        }
    }
    return field;
}
