// generator-jhipster/dist/generators/entity/command.js
import commonCommand from "../common/command.js";
import serverCommand from "../server/command.js";
const { skipCheckLengthOfIdentifier } = serverCommand.configs;
const { skipClient, skipServer } = commonCommand.configs;
const command = {
    arguments: {
        name: {
            type: String,
            required: true,
            description: 'Entity name',
        },
    },
    configs: {
        regenerate: {
            description: 'Regenerate the entity without presenting an option to update it',
            cli: {
                type: Boolean,
            },
            default: false,
            scope: 'none',
        },
        tableName: {
            description: 'Specify table name that will be used by the entity',
            cli: {
                type: String,
            },
            scope: 'none',
        },
        fluentMethods: {
            description: 'Generate fluent methods in entity beans to allow chained object construction',
            cli: {
                type: Boolean,
            },
            scope: 'none',
        },
        angularSuffix: {
            description: 'Use a suffix to generate Angular routes and files, to avoid name clashes',
            cli: {
                type: String,
            },
            scope: 'none',
        },
        clientRootFolder: {
            description: 'Use a root folder name for entities on client side. By default its empty for monoliths and name of the microservice for gateways',
            cli: {
                type: String,
            },
            scope: 'none',
        },
        skipUiGrouping: {
            description: 'Disables the UI grouping behaviour for entity client side code',
            cli: {
                type: Boolean,
            },
            scope: 'none',
        },
        skipDbChangelog: {
            description: 'Skip the generation of database changelog (liquibase for sql databases)',
            cli: {
                type: Boolean,
            },
            scope: 'none',
        },
        singleEntity: {
            description: 'Regenerate only a single entity, relationships can be not correctly generated',
            cli: {
                type: Boolean,
            },
            scope: 'none',
        },
        skipCheckLengthOfIdentifier,
        skipClient,
        skipServer,
    },
};
export default command;
