// generator-jhipster/dist/generators/client/support/types-utils.js
import { fieldTypes } from "../../../lib/jhipster/index.js";
const { BOOLEAN: TYPE_BOOLEAN, ZONED_DATE_TIME: TYPE_ZONED_DATE_TIME, INSTANT: TYPE_INSTANT, LOCAL_DATE: TYPE_LOCAL_DATE, INTEGER: TYPE_INTEGER, LONG: TYPE_LONG, BIG_DECIMAL: TYPE_BIG_DECIMAL, FLOAT: TYPE_FLOAT, DOUBLE: TYPE_DOUBLE, } = fieldTypes.CommonDBTypes;
const tsTypesByFieldType = {
    [TYPE_INTEGER]: 'number',
    [TYPE_LONG]: 'number',
    [TYPE_FLOAT]: 'number',
    [TYPE_DOUBLE]: 'number',
    [TYPE_BIG_DECIMAL]: 'number',
    [TYPE_LOCAL_DATE]: 'dayjs.Dayjs',
    [TYPE_ZONED_DATE_TIME]: 'dayjs.Dayjs',
    [TYPE_INSTANT]: 'dayjs.Dayjs',
    LocalTime: 'string',
    [TYPE_BOOLEAN]: 'boolean',
    // Enum: 'Enum',
    String: 'string',
    UUID: 'string',
    Blob: 'string',
    AnyBlob: 'string',
    TextBlob: 'string',
    ImageBlob: 'string',
    'byte[]': 'string',
    ByteBuffer: 'string',
    Duration: 'string',
};
export const fieldTsTypes = [...new Set(Object.values(tsTypesByFieldType))];
/**
 * return the input if it's a type, otherwise returns the type attribute of the input
 * @param key
 * @returns {*}
 */
export const getEntryIfTypeOrTypeAttribute = (key) => {
    if (typeof key === 'object') {
        return key.type;
    }
    return key;
};
/**
 * @private
 * Find key type for Typescript
 *
 * @param {string | object} primaryKey - primary key definition
 * @returns {string} primary key type in Typescript
 */
const getTypescriptKeyType = (primaryKey) => {
    if ([TYPE_INTEGER, TYPE_LONG, TYPE_FLOAT, TYPE_DOUBLE, TYPE_BIG_DECIMAL].includes(getEntryIfTypeOrTypeAttribute(primaryKey))) {
        return 'number';
    }
    return 'string';
};
/**
 * @private
 * Find type for Typescript
 */
export const getTypescriptType = (fieldType) => {
    return tsTypesByFieldType[fieldType];
};
export default getTypescriptKeyType;
