// generator-jhipster/dist/lib/jdl/converters/jdl-to-json/jdl-without-application-to-json-converter.js
/**
 * Copyright 2013-2026 the original author or authors from the JHipster project.
 *
 * This file is part of the JHipster project, see https://www.jhipster.tech/
 * for more information.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import BasicEntityConverter from "./jdl-to-json-basic-entity-converter.js";
import FieldConverter from "./jdl-to-json-field-converter.js";
import OptionConverter from "./jdl-to-json-option-converter.js";
import RelationshipConverter from "./jdl-to-json-relationship-converter.js";
let entities;
let jdlObject;
export default {
    convert,
};
/**
 * Converts a JDLObject to ready-to-be exported JSON entities.
 */
export function convert(jdlObjectArg, applicationName) {
    if (!jdlObjectArg || !applicationName) {
        throw new Error("The JDL object and its application's name are mandatory.");
    }
    init(jdlObjectArg);
    setBasicEntityInformation();
    setOptions();
    setFields();
    setRelationships();
    setApplicationToEntities();
    return new Map([[applicationName, Object.values(entities)]]);
}
function init(jdlObjectArg) {
    if (jdlObject) {
        resetState();
    }
    jdlObject = jdlObjectArg;
    entities = {};
}
function resetState() {
    jdlObject = null;
    entities = null;
}
function setBasicEntityInformation() {
    const convertedEntities = BasicEntityConverter.convert(jdlObject.getEntities());
    convertedEntities.forEach((jsonEntity, entityName) => {
        entities[entityName] = jsonEntity;
    });
}
function setOptions() {
    const convertedOptionContents = OptionConverter.convert(jdlObject);
    convertedOptionContents.forEach((optionContent, entityName) => {
        entities[entityName].setOptions(optionContent);
    });
}
function setFields() {
    const convertedFields = FieldConverter.convert(jdlObject);
    convertedFields.forEach((entityFields, entityName) => {
        entities[entityName].addFields(entityFields);
    });
}
function setRelationships() {
    const convertedRelationships = RelationshipConverter.convert(jdlObject.getRelationships(), jdlObject.getEntityNames());
    convertedRelationships.forEach((entityRelationships, entityName) => {
        entities[entityName].addRelationships(entityRelationships);
    });
}
function setApplicationToEntities() {
    Object.keys(entities).forEach(entityName => {
        entities[entityName].applications.push('*');
    });
}
